src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -117,10 +124,29 @@ type metricsAppenderOptions struct {
 	metrics      metricsAppenderMetrics
 }
 
-func newMetricsAppender(pool *metricsAppenderPool) *metricsAppender {
+func newMetricsAppender(
+	pool *metricsAppenderPool,
+	tagIterPool serialize.MetricTagsIteratorPool,
+	nameTag []byte) *metricsAppender {
+	tagIter := tagIterPool.Get()
 	return &metricsAppender{
 		pool:                 pool,
 		multiSamplesAppender: newMultiSamplesAppender(),
+		tagIterFn: func(tagPairs []byte) id.SortedTagIterator {
+			// Use the same tagIter for all matching computations. It is safe to reuse since a metric appender is
+			// single threaded and it's reset before each computation.
+			tagIter.Reset(tagPairs)
```

Comment:
this is the crux of the change, since it reuses the same tag iterator for all match computations in this request. previously this was fetching an iterator from the pool on every computation.